presto-hive/src/main/java/com/facebook/presto/hive/statistics/MetastoreHiveStatisticsProvider.java

```diff
@@ -246,19 +266,51 @@ private Estimate calculateNullsFraction(Map<String, PartitionStatistics> statist
                     if (partitionsWithStatisticsCount == 0) {
                         return OptionalDouble.empty();
                     }
-                    else {
-                        return OptionalDouble.of(totalPartitionsCount / partitionsWithStatisticsCount * nullsCount);
-                    }
+                    return OptionalDouble.of(totalPartitionsCount / partitionsWithStatisticsCount * nullsCount);
                 });
 
-        if (totalNullsCount.isValueUnknown() || totalRowsCount.isValueUnknown()) {
+        if (totalNullsCount.isValueUnknown()) {
             return Estimate.unknownValue();
         }
-        if (totalRowsCount.getValue() == 0.0) {
-            return Estimate.zeroValue();
+        return new Estimate(totalNullsCount.getValue() / rowCount.getValue());
+    }
+
+    private Estimate calculateDataSize(
+            Map<String, PartitionStatistics> statisticsSample,
+            String columnName,
+            Estimate rowCount,
+            Estimate nullsFraction,
+            OptionalDouble rowsPerPartition)
+    {
+        if (rowCount.isValueUnknown() || !rowsPerPartition.isPresent()) {
+            return Estimate.unknownValue();
+        }
+
+        double knownNonNullRowCount = 0;
+        double knownDataSize = 0;
+
+        for (PartitionStatistics statistics : statisticsSample.values()) {
+            double partitionRowCount = orElse(statistics.getBasicStatistics().getRowCount(), rowsPerPartition.getAsDouble());
+            HiveColumnStatistics partitionColumnStatistics = statistics.getColumnStatistics().get(columnName);
+            if (partitionColumnStatistics == null || !partitionColumnStatistics.getAverageColumnLength().isPresent()) {
+                continue;
+            }
+            double partitionNonNullCount = partitionRowCount - partitionColumnStatistics.getNullsCount().orElse(0);
+            if (partitionNonNullCount < 0) {
```

Comment:
Probably you should cap it at `0` in such case, otherwise you might end up with negative total data size estimates.

Reply:
This code is no longer there